### PR TITLE
Add tests for platform views

### PR DESCRIPTION
**Pull request by dnfield**
This adds a set of tests for iOS and Android emulator/simulator and physical devices.

Only the iOS simulator will run on CI for now - we should be able to wire these up to either devicelab or firebase testlab. Android will take a bit more work, but I'm mainly concerned about having tests for iOS that can be more easily run so I can land #11070

### CODE REVIEW

File: testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m [hunk 1]
```diff
@@ -1,12 +1,46 @@
...
+@end
 
 @implementation AppDelegate
 
 - (BOOL)application:(UIApplication*)application
     didFinishLaunchingWithOptions:(NSDictionary*)launchOptions {
   self.window = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
-  self.window.rootViewController = [[UIViewController alloc] init];
+
+  if ([[[NSProcessInfo processInfo] arguments] containsObject:@"--platform-view"]) {
```
Comment (jmagman) [inline testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m]: Can you add a comment here for when this argument would be expected?

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  NSUInteger heightA = CGImageGetHeight(imageRefA);
+  NSUInteger widthB = CGImageGetWidth(imageRefB);
+  NSUInteger heightB = CGImageGetHeight(imageRefB);
+
+  if (widthA != widthB || heightA != heightB) {
+    return NO;
+  }
+  NSUInteger bytesPerPixel = 4;
+  NSUInteger size = widthA * heightA * bytesPerPixel;
+  unsigned char* rawA = (unsigned char*)calloc(size, sizeof(unsigned char));
```
Comment (jmagman) [inline testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: There might be some better ways to compare images than comparing raw char* (some in here https://cate.blog/2015/01/21/creating-test-images-and-comparing-uiimages/ or maybe comparing UIImagePNGRepresentation?) but I'm guessing this is pretty performant as a tradeoff for being.... yeah.

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+- (NSString*)platformName {
+  NSString* simulatorName =
+      [[NSProcessInfo processInfo].environment objectForKey:@"SIMULATOR_DEVICE_NAME"];
+  if (simulatorName) {
+    return [NSString stringWithFormat:@"%@_simulator", simulatorName];
+  }
+
+  size_t size;
+  sysctlbyname("hw.model", NULL, &size, NULL, 0);
```
Comment (jmagman) [inline testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: Bleh I thought this was deprecated.
What did -[UIDevice platformString] return (I think you tried that first).

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m (hunk 2 above)
Comment (chinmaygarde) [reply testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: I we feel calloc-free is icky. `+[NSMutableData dataWithLength:]` should give you an autoreleased buffer that is zero initialized. Either way if fine but that might make it harder to forget to free these buffers on early returns.

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+  CGContextRef contextB =
+      CGBitmapContextCreate(rawB, widthA, heightA, bitsPerComponent, bytesPerRow, colorSpace,
+                            kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
+  CGColorSpaceRelease(colorSpace);
+
+  CGContextDrawImage(contextB, CGRectMake(0, 0, widthA, heightA), imageRefB);
+  CGContextRelease(contextB);
+
+  for (int i = 0; i < size; ++i) {
```
Comment (chinmaygarde) [inline testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: `memcmp`?

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m (hunk 2 above)
Comment (dnfield) [reply testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: I don't want to compare the PNG bytes - I was getting false negatives that way in local tests, and I suspect that different versions of the OS/devices will end up decoding or encoding slightly differently.  If there's a better way to get the raw pixels I'm happy to do that.

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m (hunk 3 above)
Comment (dnfield) [reply testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: That was a category someone made that delegates to this, more or less.  This is probably not safe for stores but I only need it for tests.

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m (hunk 2 above)
Comment (dnfield) [reply testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: How would I pass the NSMutableData in?

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m (hunk 4 above)
Comment (dnfield) [reply testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: D'oh.

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m (hunk 2 above)
Comment (dnfield) [reply testing/scenario_app/ios/Scenarios/ScenariosUITests/PlatformViewUITests.m]: Oh I see..  I'll do that instead, definitely better.

File: testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m (hunk 1 above)
Comment (dnfield) [reply testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m]: Done

**Comment by dnfield** [comment]
Going to land this to kick infra, since it seems like a safe, test only change.

@jmagman if there's anything else here you wanted to discuss I'm happy to follow up.

**Pull request merged**